File: recursion/swap-nodes-in-pairs.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def swapPairs(self, head):
        """
        :type head: Optional[ListNode]
        :rtype: Optional[ListNode]
        """
        if(head == None or head.next == None):
            return head
        
        f = head
        s = head.next
        p = None

        while(f != None and s != None):
            t = s.next
            s.next = f
            f.next = t
            if(p != None):
                p.next = s
            else:
                head = s
            p = f
            f = t
            if(t != None):
                s = t.next
            else:
                s = None
        
        return head

File: linked-list/swap-nodes-in-pairs.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        if(head == nullptr || head->next == nullptr){
            return head;
        }
        ListNode* temp = head;
        ListNode* prev = nullptr;
        while(temp != nullptr && temp->next != nullptr){
            ListNode* temp1 = temp->next;
            temp->next = temp1->next;
            temp1->next = temp;
            if(temp == head){
                head = temp1;
            }
            if(prev != nullptr){
                prev->next = temp1;
            }
            prev = temp;
            temp = temp->next;
        }
        return head;
    }
};